acme/tf/networks/legal_actions.py

```diff
@@ -0,0 +1,119 @@
+# python3
+# Copyright 2018 DeepMind Technologies Limited. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Networks used for handling illegal actions."""
+
+from typing import Any, Callable, Iterable, Optional, Text, Union
+
+import acme
+# pytype: disable=import-error
+from acme.wrappers import open_spiel_wrapper
+# pytype: enable=import-error
+import numpy as np
+import sonnet as snt
+import tensorflow as tf
+import tensorflow_probability as tfp
+
+tfd = tfp.distributions
+
+
+class MaskedSequential(snt.Module):
+  """Applies a legal actions mask to a linear chain of modules / callables.
+
+  It is assumed the trailing dimension of the final layer (representing
+  action values) is the same as the trailing dimension of legal_actions.
+  """
+
+  def __init__(self,
+               layers: Iterable[Callable[..., Any]] = None,
+               name: Optional[Text] = None):
+    super(MaskedSequential, self).__init__(name=name)
+    self._layers = list(layers) if layers is not None else []
+    self._illegal_action_penalty = -1e9
+    # Note: illegal_action_penalty cannot be -np.inf, throws error "Priority
```

Comment:
This is interesting, did this surprise you? Is the illegal action being used to calculate the priority somehow?

Reply:
Got to the bottom of this. Trfl’s `qlearning` ops use a `batched_index` function to return the action value of selected actions: `qa_tm1 = indexing_ops.batched_index(q_tm1, a_tm1)`

And `batched_index` computes: `tf.reduce_sum(values * one_hot_indices, axis=-1, keepdims=keepdims)`

This returns NaN whenever `-np.inf` is present among the actions values. These NaNs show up in the td_error, which then gets passed to the priorities, which throws the error. 

Note: this examination prompted the following commit fa986b20be5244dcc72b49bcbda50dbe4adf0cf3 allowing the MaskedSequential module to handle `-np.inf` correctly. My first version of MaskedSequential did handle `-np.inf` correctly, but ran into the priority error, so I switched to using `-1e9`. Then I subsequently added this line of code: `outputs = outputs * (1 - terminal)` which seemed cleaner than what was previously there, but broke the ability to handle `-np.inf`. Either way, whether `-np.inf` is used or not, I think the update in this commit is the correct way to force the network to output all zeros in a terminal state.

**TLDR**: Ultimately, outputting `-np.inf` for illegal actions is incompatible with trfl. There was also an issue with how MaskedSequential handled `-np.inf` which is now fixed.